File: AIAgent/ZBuild/Repos/cxdb-main/clients/go/context.go
// Copyright 2025 StrongDM Inc
// SPDX-License-Identifier: Apache-2.0

package cxdb

import (
	"context"
	"encoding/binary"
	"fmt"
)

// ContextHead represents the head of a context (branch).
type ContextHead struct {
	ContextID  uint64
	HeadTurnID uint64
	HeadDepth  uint32
}

// CreateContext creates a new context in CXDB.
// If baseTurnID is 0, creates an empty context.
// If baseTurnID is non-zero, creates a context starting from that turn.
func (c *Client) CreateContext(ctx context.Context, baseTurnID uint64) (*ContextHead, error) {
	payload := make([]byte, 8)
	binary.LittleEndian.PutUint64(payload, baseTurnID)

	resp, err := c.sendRequest(ctx, msgCtxCreate, payload)
	if err != nil {
		return nil, fmt.Errorf("create context: %w", err)
	}

	return parseContextHead(resp.payload)
}

// ForkContext creates a new context branching from a specific turn.
// This is an O(1) operation - it creates a new head pointer without copying data.
func (c *Client) ForkContext(ctx context.Context, baseTurnID uint64) (*ContextHead, error) {
	payload := make([]byte, 8)
	binary.LittleEndian.PutUint64(payload, baseTurnID)

	resp, err := c.sendRequest(ctx, msgCtxFork, payload)
	if err != nil {
		return nil, fmt.Errorf("fork context: %w", err)
	}

	return parseContextHead(resp.payload)
}

// GetHead retrieves the current head of a context.
func (c *Client) GetHead(ctx context.Context, contextID uint64) (*ContextHead, error) {
	payload := make([]byte, 8)
	binary.LittleEndian.PutUint64(payload, contextID)

	resp, err := c.sendRequest(ctx, msgGetHead, payload)
	if err != nil {
		return nil, fmt.Errorf("get head: %w", err)
	}

	return parseContextHead(resp.payload)
}

func parseContextHead(payload []byte) (*ContextHead, error) {
	if len(payload) < 20 {
		return nil, fmt.Errorf("%w: context head too short (%d bytes)", ErrInvalidResponse, len(payload))
	}
	return &ContextHead{
		ContextID:  binary.LittleEndian.Uint64(payload[0:8]),
		HeadTurnID: binary.LittleEndian.Uint64(payload[8:16]),
		HeadDepth:  binary.LittleEndian.Uint32(payload[16:20]),
	}, nil
}

File: AIAgent/ZBuild/Repos/cxdb-main/clients/rust/src/context.rs
// Copyright 2025 StrongDM Inc
// SPDX-License-Identifier: Apache-2.0

use byteorder::{LittleEndian, ReadBytesExt, WriteBytesExt};

use crate::client::{Client, RequestContext};
use crate::error::{Error, Result};
use crate::protocol::{MSG_CTX_CREATE, MSG_CTX_FORK, MSG_GET_HEAD};

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct ContextHead {
    pub context_id: u64,
    pub head_turn_id: u64,
    pub head_depth: u32,
}

impl Client {
    pub fn create_context(&self, ctx: &RequestContext, base_turn_id: u64) -> Result<ContextHead> {
        let mut payload = Vec::with_capacity(8);
        payload.write_u64::<LittleEndian>(base_turn_id)?;
        let frame = self.send_request(ctx, MSG_CTX_CREATE, &payload)?;
        parse_context_head(&frame.payload)
    }

    pub fn fork_context(&self, ctx: &RequestContext, base_turn_id: u64) -> Result<ContextHead> {
        let mut payload = Vec::with_capacity(8);
        payload.write_u64::<LittleEndian>(base_turn_id)?;
        let frame = self.send_request(ctx, MSG_CTX_FORK, &payload)?;
        parse_context_head(&frame.payload)
    }

    pub fn get_head(&self, ctx: &RequestContext, context_id: u64) -> Result<ContextHead> {
        let mut payload = Vec::with_capacity(8);
        payload.write_u64::<LittleEndian>(context_id)?;
        let frame = self.send_request(ctx, MSG_GET_HEAD, &payload)?;
        parse_context_head(&frame.payload)
    }
}

fn parse_context_head(payload: &[u8]) -> Result<ContextHead> {
    if payload.len() < 20 {
        return Err(Error::invalid_response(format!(
            "context head too short ({} bytes)",
            payload.len()
        )));
    }
    let mut cursor = std::io::Cursor::new(payload);
    Ok(ContextHead {
        context_id: cursor.read_u64::<LittleEndian>()?,
        head_turn_id: cursor.read_u64::<LittleEndian>()?,
        head_depth: cursor.read_u32::<LittleEndian>()?,
    })
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::test_util::{decode_hex, load_fixture};

    fn payload_u64(value: u64) -> Vec<u8> {
        let mut payload = Vec::with_capacity(8);
        payload.write_u64::<LittleEndian>(value).unwrap();
        payload
    }

    #[test]
    fn context_payloads_match_fixtures() {
        let fixture = load_fixture("ctx_create_base0");
        assert_eq!(fixture.msg_type, MSG_CTX_CREATE);
        assert_eq!(decode_hex(&fixture.payload_hex), payload_u64(0));

        let fixture = load_fixture("ctx_fork_base123");
        assert_eq!(fixture.msg_type, MSG_CTX_FORK);
        assert_eq!(decode_hex(&fixture.payload_hex), payload_u64(123));

        let fixture = load_fixture("get_head_ctx42");
        assert_eq!(fixture.msg_type, MSG_GET_HEAD);
        assert_eq!(decode_hex(&fixture.payload_hex), payload_u64(42));
    }
}
